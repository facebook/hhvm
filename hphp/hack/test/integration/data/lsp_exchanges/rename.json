[
  {
    "jsonrpc": "2.0",
    "id": 990,
    "method": "initialize",
    "params": {
      "initializationOptions": {},
      "processId": null,
      "rootPath": "${root_path}",
      "capabilities": {}
    }
  },
  {
    "jsonrpc": "2.0",
    "method": "textDocument/didOpen",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}",
        "languageId": "hack",
        "version": 1,
        "text": "${php_file}"
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 1,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 9,
        "character": 42
      },
      "newName": "NEW_CONSTANT"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 2,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 2,
        "character": 11
      },
      "newName": "HistoryClass"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 3,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 51,
        "character": 18
      },
      "newName": "some_other_method"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 4,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 53,
        "character": 34
      },
      "newName": "ANOTHER_STR_CONSTANT"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 5,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 54,
        "character": 39
      },
      "newName": "AVOGADRO_CONSTANT"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 6,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 73,
        "character": 23
      },
      "newName": "kittens_and_rainbows"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 7,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 56,
        "character": 29
      },
      "newName": "new_and_improved"
    }
  },
  {
    "jsonrpc": "2.0",
     "id": 8,
     "method": "textDocument/rename",
     "params": {
       "textDocument": {
         "uri": "${php_file_uri}"
       },
       "position": {
         "line": 26,
         "character": 33
       },
       "newName": "new_static_method"
     }
   },
   {
     "jsonrpc": "2.0",
     "id": 9,
     "method": "textDocument/rename",
     "params": {
       "textDocument": {
         "uri": "${php_file_uri}"
       },
       "position": {
         "line": 31,
         "character": 33
       },
       "newName": "new_async_method"
     }
   },
  {
    "jsonrpc": "2.0",
     "id": 10,
     "method": "textDocument/rename",
     "params": {
       "textDocument": {
         "uri": "${php_file_uri}"
       },
       "position": {
         "line": 36,
         "character": 42
       },
       "newName": "new_static_async_method"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 11,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 65,
        "character": 11
      },
      "newName": "new_test_rename_method"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 12,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 66,
        "character": 29
      },
      "newName": "new_async_test_rename_method"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 13,
    "method": "textDocument/rename",
    "params": {
      "textDocument": {
        "uri": "${php_file_uri}"
      },
      "position": {
        "line": 57,
        "character": 13
      },
      "newName": "TSHIRT_SIZES"
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 999,
    "method": "shutdown",
    "params": {}
  }
]
