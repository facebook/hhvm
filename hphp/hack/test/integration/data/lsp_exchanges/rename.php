<?hh //strict

class TestClass { // 2. Rename TestClass
  const CONSTANT = 5;
  const string STR_CONSTANT = "hello";
  const vec<int> VEC_INT_CONSTANT = vec [1, 2, 3, 4];
  public int $property;

  public function __construct(int $i) {
    $this->property = $i + TestClass::CONSTANT; // 1. Rename CONSTANT
  }

  public function test_method(): int {
    return $this->property;
  }

  /**
   * Some docblock
   *
   *
   */
  public function deprecated_method(int $x, string $y, int $v, ...): int {
    return $x;
  }

  // 8. Rename depr_static
  public static function depr_static(int $x, string $y, int $v, ...): int {
    return $x;
  }

  // 9. Rename depr_async
  public async function depr_async(int $x, int $v, ...): Awaitable<int> {
    return $x;
  }

  // 10. Rename depr_static_async
  public static async function depr_static_async(int $x): Awaitable<
    void,
  >
  {}
}

enum RenameTestEnum: int {
  SMALL = 0;
  MEDIUM = 1;
  LARGE = 2;
}

function test_rename(): void {
  $test_class = new TestClass(1);
  // 3. Rename 1st test_method
  $test_class->test_method(); $test_class->test_method();
  $const = TestClass::CONSTANT;
  $str_const = TestClass::STR_CONSTANT; // 4. Rename STR_CONSTANT
  $vec_int_const = TestClass::VEC_INT_CONSTANT; // 5. Rename VEC_INT_CONSTANT
  // 7. Renaming deprecated_method
  $num = $test_class->deprecated_method(5, "", 4, 4, 3, 5, "hello");
  $size = RenameTestEnum::SMALL; // 13. Rename TestEnum
}

async function async_test_rename(): Awaitable<int> {
  return 5;
}

async function call_test_rename(): Awaitable<int> {
  test_rename(); // 11. Rename test_rename
  return await async_test_rename(); // 12. Rename async_test_rename
}

function test_rename_localvar(int $local): void {      //  Should match
  $local = 3;                                           //  Should match
  j($local) + $local + h("\$x = $local");     //  1st, 2nd, and 4th should match
  $lambda1 = $x ==> $x + 1;                         //  Should not match
  $lambda2 = $a ==> $local + $a; // 6. Renaming this $local    //  Should match
  $lambda3 = function($x) {                         //  Should not match
    return $x + 1; };                               //  Should not match
  $lambda4 = function($b) use($local) {                 //  Should match
    return $local + $b; };                              //  Should match
}

type RenameShapeKeyEnum =
  shape(RenameTestEnum::SMALL => int); // 4. Rename STR_CONSTANT

type RenameShapeKeyConstant =
  shape(TestClass::STR_CONSTANT => int); // 4. Rename STR_CONSTANT

function test_rename_shape_keys(): void {
  shape(RenameTestEnum::SMALL => 123); // 4. Rename STR_CONSTANT
  shape(TestClass::STR_CONSTANT => 123); // 13. Rename TestEnum
}
