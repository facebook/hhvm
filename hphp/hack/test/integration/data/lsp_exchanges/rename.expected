[
  {
    "jsonrpc": "2.0",
    "id": 990,
    "result": {
      "capabilities": {
        "textDocumentSync": {
          "openClose": true,
          "change": 2,
          "willSave": false,
          "willSaveWaitUntil": false,
          "save": {
            "includeText": false
          }
        },
        "hoverProvider": true,
        "completionProvider": {
          "resolveProvider": true,
          "triggerCharacters": [
            "$",
            ">",
            "\\",
            ":",
            "<",
            "[",
            "'",
            "\""
          ]
        },
        "signatureHelpProvider": {
          "triggerCharacters": [
            "(",
            ","
          ]
        },
        "definitionProvider": true,
        "referencesProvider": true,
        "documentHighlightProvider": true,
        "documentSymbolProvider": true,
        "workspaceSymbolProvider": true,
        "codeActionProvider": false,
        "documentFormattingProvider": true,
        "documentRangeFormattingProvider": true,
        "documentOnTypeFormattingProvider": {
          "firstTriggerCharacter": ";",
          "moreTriggerCharacter": [
            "}"
          ]
        },
        "renameProvider": true,
        "typeCoverageProvider": true,
        "rageProvider": true
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 3,
                "character": 8
              },
              "end": {
                "line": 3,
                "character": 16
              }
            },
            "newText": "NEW_CONSTANT"
          },
          {
            "range": {
              "start": {
                "line": 52,
                "character": 22
              },
              "end": {
                "line": 52,
                "character": 30
              }
            },
            "newText": "NEW_CONSTANT"
          },
          {
            "range": {
              "start": {
                "line": 9,
                "character": 38
              },
              "end": {
                "line": 9,
                "character": 46
              }
            },
            "newText": "NEW_CONSTANT"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 2,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 2,
                "character": 6
              },
              "end": {
                "line": 2,
                "character": 15
              }
            },
            "newText": "HistoryClass"
          },
          {
            "range": {
              "start": {
                "line": 54,
                "character": 19
              },
              "end": {
                "line": 54,
                "character": 28
              }
            },
            "newText": "HistoryClass"
          },
          {
            "range": {
              "start": {
                "line": 53,
                "character": 15
              },
              "end": {
                "line": 53,
                "character": 24
              }
            },
            "newText": "HistoryClass"
          },
          {
            "range": {
              "start": {
                "line": 52,
                "character": 11
              },
              "end": {
                "line": 52,
                "character": 20
              }
            },
            "newText": "HistoryClass"
          },
          {
            "range": {
              "start": {
                "line": 49,
                "character": 20
              },
              "end": {
                "line": 49,
                "character": 29
              }
            },
            "newText": "HistoryClass"
          },
          {
            "range": {
              "start": {
                "line": 9,
                "character": 27
              },
              "end": {
                "line": 9,
                "character": 36
              }
            },
            "newText": "HistoryClass"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 3,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 12,
                "character": 18
              },
              "end": {
                "line": 12,
                "character": 29
              }
            },
            "newText": "some_other_method"
          },
          {
            "range": {
              "start": {
                "line": 51,
                "character": 43
              },
              "end": {
                "line": 51,
                "character": 54
              }
            },
            "newText": "some_other_method"
          },
          {
            "range": {
              "start": {
                "line": 51,
                "character": 15
              },
              "end": {
                "line": 51,
                "character": 26
              }
            },
            "newText": "some_other_method"
          },
          {
            "range": {
              "start": {
                "line": 11,
                "character": 0
              },
              "end": {
                "line": 11,
                "character": 0
              }
            },
            "newText": "\n  <<__Deprecated(\"Use `some_other_method` instead\")>>\n  public function test_method(): int {\n    return $this->some_other_method();\n  }\n"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 4,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 4,
                "character": 15
              },
              "end": {
                "line": 4,
                "character": 27
              }
            },
            "newText": "ANOTHER_STR_CONSTANT"
          },
          {
            "range": {
              "start": {
                "line": 53,
                "character": 26
              },
              "end": {
                "line": 53,
                "character": 38
              }
            },
            "newText": "ANOTHER_STR_CONSTANT"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 5,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 5,
                "character": 17
              },
              "end": {
                "line": 5,
                "character": 33
              }
            },
            "newText": "AVOGADRO_CONSTANT"
          },
          {
            "range": {
              "start": {
                "line": 54,
                "character": 30
              },
              "end": {
                "line": 54,
                "character": 46
              }
            },
            "newText": "AVOGADRO_CONSTANT"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 6,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 69,
                "character": 34
              },
              "end": {
                "line": 69,
                "character": 40
              }
            },
            "newText": "$kittens_and_rainbows"
          },
          {
            "range": {
              "start": {
                "line": 70,
                "character": 2
              },
              "end": {
                "line": 70,
                "character": 8
              }
            },
            "newText": "$kittens_and_rainbows"
          },
          {
            "range": {
              "start": {
                "line": 71,
                "character": 4
              },
              "end": {
                "line": 71,
                "character": 10
              }
            },
            "newText": "$kittens_and_rainbows"
          },
          {
            "range": {
              "start": {
                "line": 71,
                "character": 14
              },
              "end": {
                "line": 71,
                "character": 20
              }
            },
            "newText": "$kittens_and_rainbows"
          },
          {
            "range": {
              "start": {
                "line": 71,
                "character": 32
              },
              "end": {
                "line": 71,
                "character": 38
              }
            },
            "newText": "$kittens_and_rainbows"
          },
          {
            "range": {
              "start": {
                "line": 73,
                "character": 20
              },
              "end": {
                "line": 73,
                "character": 26
              }
            },
            "newText": "$kittens_and_rainbows"
          },
          {
            "range": {
              "start": {
                "line": 76,
                "character": 30
              },
              "end": {
                "line": 76,
                "character": 36
              }
            },
            "newText": "$kittens_and_rainbows"
          },
          {
            "range": {
              "start": {
                "line": 77,
                "character": 11
              },
              "end": {
                "line": 77,
                "character": 17
              }
            },
            "newText": "$kittens_and_rainbows"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 7,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 21,
                "character": 18
              },
              "end": {
                "line": 21,
                "character": 35
              }
            },
            "newText": "new_and_improved"
          },
          {
            "range": {
              "start": {
                "line": 56,
                "character": 22
              },
              "end": {
                "line": 56,
                "character": 39
              }
            },
            "newText": "new_and_improved"
          },
          {
            "range": {
              "start": {
                "line": 15,
                "character": 0
              },
              "end": {
                "line": 15,
                "character": 0
              }
            },
            "newText": "\n  <<__Deprecated(\"Use `new_and_improved` instead\")>>\n  public function deprecated_method(int $x, string $y, int $v, ...): int {\n    return $this->new_and_improved($x, $y, $v, ...Vec\\drop(func_get_args(), 3));\n  }\n"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 8,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 26,
                "character": 25
              },
              "end": {
                "line": 26,
                "character": 36
              }
            },
            "newText": "new_static_method"
          },
          {
            "range": {
              "start": {
                "line": 24,
                "character": 0
              },
              "end": {
                "line": 24,
                "character": 0
              }
            },
            "newText": "\n  <<__Deprecated(\"Use `new_static_method` instead\")>>\n  public static function depr_static(int $x, string $y, int $v, ...): int {\n    return self::new_static_method($x, $y, $v, ...Vec\\drop(func_get_args(), 3));\n  }\n"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 9,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 31,
                "character": 24
              },
              "end": {
                "line": 31,
                "character": 34
              }
            },
            "newText": "new_async_method"
          },
          {
            "range": {
              "start": {
                "line": 29,
                "character": 0
              },
              "end": {
                "line": 29,
                "character": 0
              }
            },
            "newText": "\n  <<__Deprecated(\"Use `new_async_method` instead\")>>\n  public async function depr_async(int $x, int $v, ...): Awaitable<int> {\n    return await $this->new_async_method($x, $v, ...Vec\\drop(func_get_args(), 2));\n  }\n"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 10,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 36,
                "character": 31
              },
              "end": {
                "line": 36,
                "character": 48
              }
            },
            "newText": "new_static_async_method"
          },
          {
            "range": {
              "start": {
                "line": 34,
                "character": 0
              },
              "end": {
                "line": 34,
                "character": 0
              }
            },
            "newText": "\n  <<__Deprecated(\"Use `new_static_async_method` instead\")>>\n  public static async function depr_static_async(int $x): Awaitable<\n    void,\n  > {\n    await self::new_static_async_method($x);\n  }\n"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 11,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 48,
                "character": 9
              },
              "end": {
                "line": 48,
                "character": 20
              }
            },
            "newText": "new_test_rename_method"
          },
          {
            "range": {
              "start": {
                "line": 65,
                "character": 2
              },
              "end": {
                "line": 65,
                "character": 13
              }
            },
            "newText": "new_test_rename_method"
          },
          {
            "range": {
              "start": {
                "line": 47,
                "character": 0
              },
              "end": {
                "line": 47,
                "character": 0
              }
            },
            "newText": "\n<<__Deprecated(\"Use `new_test_rename_method` instead\")>>\nfunction test_rename(): void {\n  new_test_rename_method();\n}\n"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 12,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 60,
                "character": 15
              },
              "end": {
                "line": 60,
                "character": 32
              }
            },
            "newText": "new_async_test_rename_method"
          },
          {
            "range": {
              "start": {
                "line": 66,
                "character": 15
              },
              "end": {
                "line": 66,
                "character": 32
              }
            },
            "newText": "new_async_test_rename_method"
          },
          {
            "range": {
              "start": {
                "line": 59,
                "character": 0
              },
              "end": {
                "line": 59,
                "character": 0
              }
            },
            "newText": "\n<<__Deprecated(\"Use `new_async_test_rename_method` instead\")>>\nasync function async_test_rename(): Awaitable<int> {\n  return await new_async_test_rename_method();\n}\n"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 13,
    "result": {
      "changes": {
        "${php_file_uri}": [
          {
            "range": {
              "start": {
                "line": 42,
                "character": 5
              },
              "end": {
                "line": 42,
                "character": 19
              }
            },
            "newText": "TSHIRT_SIZES"
          },
          {
            "range": {
              "start": {
                "line": 57,
                "character": 10
              },
              "end": {
                "line": 57,
                "character": 24
              }
            },
            "newText": "TSHIRT_SIZES"
          }
        ]
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 999,
    "result": null
  }
]
